consensus/istanbul/backend/engine_test.go

```diff
@@ -506,3 +577,175 @@ func TestWriteCommittedSeals(t *testing.T) {
+		isIstanbulCompatible  bool     // whether or not if the inserted block is istanbul compatible
+		expectedValidatorsNum int      // the number of expected validators
+		expectedDemotedNum    int      // the number of expected demoted validators
+	}
+
+	testcases := []testcase{
+		// The following testcases are the ones before istanbul incompatible change
+		{
+			[]uint64{5000000, 5000000, 5000000, 5000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 5000000, 5000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 5000000, 6000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 6000000, 6000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{6000000, 6000000, 6000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		// The following testcases are the ones after istanbul incompatible change
+		{
+			[]uint64{5000000, 5000000, 5000000, 5000000},
+			true,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 5000000, 5000000, 6000000},
+			true,
+			1,
+			3,
+		},
+		{
+			[]uint64{5000000, 5000000, 6000000, 6000000},
+			true,
+			2,
+			2,
+		},
+		{
+			[]uint64{5000000, 6000000, 6000000, 6000000},
+			true,
+			3,
+			1,
+		},
+		{
+			[]uint64{6000000, 6000000, 6000000, 6000000},
+			true,
+			4,
+			0,
+		},
+		{
+			[]uint64{5500001, 5500000, 5499999, 0},
+			true,
+			2,
+			2,
+		},
+	}
+
+	testNum := 4
+	ms := uint64(5500000)
+	configItems := makeSnapshotTestConfigItems()
+	configItems = append(configItems, minimumStake(new(big.Int).SetUint64(ms)))
+	for _, tc := range testcases {
+		if tc.isIstanbulCompatible {
+			configItems = append(configItems, istanbulCompatibleBlock(new(big.Int).SetUint64(0)))
+		}
+		chain, engine := newBlockChain(testNum, configItems...)
+
+		// set old staking manager after finishing this test.
+		oldStakingManager := reward.GetStakingManager()
+
+		// set new staking manager with the given staking information.
+		stakingInfo := makeFakeStakingInfo(0, nodeKeys, tc.stakingAmounts)
+		reward.SetTestStakingManagerWithStakingInfoCache(stakingInfo)
+
+		block := makeBlockWithSeal(chain, engine, chain.Genesis())
+		_, err := chain.InsertChain(types.Blocks{block})
+		assert.NoError(t, err)
+
+		snap, err := engine.snapshot(chain, block.NumberU64(), block.Hash(), nil)
+		assert.NoError(t, err)
+		assert.Equal(t, tc.expectedValidatorsNum, len(snap.ValSet.List()))
+		assert.Equal(t, tc.expectedDemotedNum, len(snap.ValSet.DemotedList()))
+
+		if tc.isIstanbulCompatible && !isAllDemoted(ms, tc.stakingAmounts) {
+			// if the inserted block is istanbul compatible
+			// and a validator has enough KLAYs at least
+			for idx, sa := range tc.stakingAmounts {
+				if sa >= ms {
+					assert.True(t, checkInValidators(addrs[idx], snap.ValSet.List()))
+				} else {
+					assert.True(t, checkInValidators(addrs[idx], snap.ValSet.DemotedList()))
+				}
+			}
+		} else {
+			// if the inserted block is not istanbul compatible
+			// and all validators doesn't have enough KLAYs
```

Comment:
```suggestion
			// or all validators don't have enough KLAYs
```
Isn't it `or` if above `if` statement does not meet the condition?

Reply:
You are right. Reflected :)